README.md
<p align="center"><img src="https://wbmtranslator.esaracco.fr/images/wbmtranslator.png"/></p>

# wbmtranslator

 [Project homepage](https://wbmtranslator.esaracco.fr)

wbmtranslator is the translation assistant for [Webmin](http://www.webmin.com) / [Usermin](http://www.webmin.com/usermin.html) modules. It allows Webmin / Usermin translators to have a global view of current translations for all registered modules and themes, including the web interface and configuration screens. They can accurately track changes, create or update translations, send them to the Webmin team, and more.

## Installation

### From a wbm archive:

Open the Webmin modules manager and upload the wbmtranslator file.

### From the Git repository:

1. Rename your local Git repository (optional):
```bash
mv wbmtranslator/ translator/
```
2. Build a gzipped tarball archive:
```bash
tar zcvf wbmtranslator.wbm.gz translator/
```
3. Open the Webmin modules manager and upload your brand new wbmtranslator file.

## Perl dependencies

You need at least the following Perl modules in order to run wbmtranslator:

- Date::Manip
- Digest::MD5
- Encode
- File::Basename
- File::Copy
- File::Path
- HTML::Entities
- JSON
- LWP::UserAgent
- MIME::Lite
- POSIX

Depending on the software installed on your system and your wbmtranslator configuration options, you may also need the following modules:

- Authen::NTLM
- Digest::HMAC_MD5
- Mail::Sender

All those modules can be loaded from CPAN.

As root you can try:

```bash
# perl -MCPAN -e shell
CPAN Shell> install module1 module2 ...
```

## License
GPL


ChangeLog
2021-01-24 Emmanuel Saracco <emmanuel@esaracco.fr>

	* Version: 0.9.2

	* Transation:

	  - UTF-8 is the webmin default now.

2021-01-22 Emmanuel Saracco <emmanuel@esaracco.fr>

	* Version: 0.9.1

	  - Fixed issue with Reverso translation API.

2019-08-01 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.9

	  - Big rewrite of the entire module in order to take full advantage
	    of the power of the Authentic theme.
	  - Removed no longer maintained translations.
	  - Added a configuration option to check the release of new versions
	    of wbmtranslator.
	  - Added tabs for "Translation management" and "Informations
	    translation" sections.
	  - Shortening the labels for the icons on the main page.
	  - Added help for each feature.
	  - As usual: a lot of debugging.

2019-06-29 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.8.1

	* Help translation section:

	  - Removed unused feature.

	* Other:

	  - Logo update.
	  - Project links update.
	  - HTML fixes and code cleaning.

2019-06-05 Emmanuel Saracco <emmanuel@esaracco.fr>

	* wbmtranslator is now hosted on GitHub.

2019-05-26 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.8

	* Configuration:

	  - Updated webmin and Usermin core modules list.

	* Translation:

	  - Fixed a lot of problems with translation management and UTF-8.
	    Please backup data before using this new release, and send me a
	    email if things go wrong...

	  - Fixed problems with ref file items update detection.

	* Translation administration section:

	  - Added purge of unused items in ref file.

	* Translation console:

	  - Fixed Reverso translation console, so it works again, and much
	    better before.

2014-07-31 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.7.3

	* Configuration:

	  - Updated "*-core-modules.data".

2013-03-20 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* send translations section:
	
	  - Increased combobox from 5 to 10 entries.

2013-01-04 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.7.2

	* Configuration:

	  - Updated "webmin-core-modules.data".

2012-11-16 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Translation administration section:

	  - Fixed a bug with Webmin translations.

	* Translation console:

	  - Fixed encoding.
	  - Save/restore user translation choice in/from a cookie.

2012-11-13 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Send translations section:

	  - Fixed a bug with Webmin translations.

2012-11-12 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Translation console:

	  - Removed Bing translation console and replaced it with Reverso.

	* Configuration:

	  - Updated "webmin-core-modules.data". Moved *modules.data files to
	    "data/" directory.

2012-11-09 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Translation console:

	  - Removed Google translation console and replace it with Bing.

2011-07-27 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Other:

	  - Updated wbmtranslator download URL.

2009-12-25 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.7.1

	* Send translations section:
	
	  - Fixed a problem with email validation.

2009-12-23 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Translations administration section:

	  - Fixed a problem with UTF-8 detection. Now all languages appear
	    in lists, and UTF-8 translation can be done.
	

2009-12-06 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.7

	* Translations archives:

	  - Now include all needed Usermin files.

	* Configuration:

	  - Fixed a problem in core modules list updater script.
	  - Updated Webmin core modules list.

2009-07-23 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Configuration:

	  - Added "trans_webmin" and "trans_working_path" option in a new
	    "line_global" section to avoid buggy themes that were not display
	    them in the configuration page.

2009-07-21 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Other:
	
	  - Added support for themes translation (task #1907).

	* Module information translation section:

	  - Added support for long description translation (bug #1908).


	* Google translation console:

	  - Always use UTF-8 for translator assistant. Waiting for translators
	    feedback!

	* Configuration:

		- Updated webmin and Usermin core modules list.

2008-03-29 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.6.3

	* Google translation console:

	  - Fixed a lot of encoding problems.
	  - Added new languages (Greek and Dutch)

	* Send translations section:

	  - Added translator name information.
	  - Updated E-Mail content to Webmin translations team to be more 
	    consistent regarding to the translation type (Webmin ou Usermin).

	* Configuration:
	  - Added variable "trans_name" in "config.info" file.
	  - Updated webmin core modules list.

2008-03-27 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Informations translation:

	  - Fixed a problem with "module.info" translation when it does not
	    have all descriptions for all module's languages.

2007-10-06 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.6.2

	* Configuration:
		- Updated webmin and Usermin core modules list.

	* Other:
		- Fixed a problem with the "Google translation console". It
		  work again now :-) Added new languages.

2006-06-18 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Other:
		- Fixed a problem's assumption with module name.

2006-05-23 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.6.1

	* Configuration:
		- Updated webmin core modules list.

2006-05-21 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Translations archives:

js/scripts.js
function HTMLClassRemove (e, cls)
{
  e.className = e.className.replace (cls, '');
}

function HTMLClassAdd (e, cls)
{
  if (e.className.indexOf (cls) == -1)
    e.className += ' '+cls;
}

function HTMLClassReplace (e, cls1, cls2)
{
  HTMLClassRemove (e, cls1);
  HTMLClassAdd (e, cls2);
}

function updateActionsChecked (form, target, name)
{
  for (var i = 0, len = form.elements.length; i < len; i++)
  {
    var e = form.elements[i];

    if (e.type == 'checkbox' && e.checked && e.name.indexOf(name) != -1)
      return HTMLClassRemove (target, 'disabled');
  }

  HTMLClassAdd (target, 'disabled');
}

function checkUncheckAll (e, name, checkedText, uncheckedText)
{
  var r = document.querySelectorAll ('input[name^="'+name+'"]');
  var v = e.getAttribute ('data-checked');

  for (var i = 0, len = r.length; i < len; i++)
    r[i].checked = (v == 0);

  if (v == 1)
  {
    e.setAttribute ('data-checked', 0);
    HTMLClassReplace (e.querySelector('i'), 'fa-square-o', 'fa-check-square-o');
    e.querySelector('span').innerText = checkedText;
  }
  else
  {
    e.setAttribute ('data-checked', 1);
    HTMLClassReplace (e.querySelector('i'), 'fa-check-square-o', 'fa-square-o');
    e.querySelector('span').innerText = uncheckedText;
  }
}

function displayMsg (msg, type, title)
{
  var e = document.getElementById ('trans-msg');

  e.innerHTML =
    "<button class='close' onclick='document.getElementById(\"trans-msg\").style.display=\"none\"'>&times;</button>"+((title)?"<strong>"+title+"</strong>":'')+"<p>"+msg+"</p>";

  e.className = 'alert alert-'+((type)?type:'info');
  e.style.display = 'block';
}


function translate_console_open (lang2)
{
  lang2 = (lang2) ? lang2.replace (/^([a-z]+).*/i, "$1") : '';

  window.open(
    'translate_popup.cgi?lang2='+lang2,
    'translate_popup',
    'toolbar=no,location=no,directories=no,status=no,scrollbars=yes,' +
    'resizable=no,copyhistory=no,width=600,height=400'
  );
}


js/translator.js
function HTMLClassRemove (e, cls)
{
  e.className = e.className.replace (cls, '');
}

function HTMLClassAdd (e, cls)
{
  if (e.className.indexOf (cls) == -1)
    e.className += ' '+cls;
}

function HTMLClassReplace (e, cls1, cls2)
{
  HTMLClassRemove (e, cls1);
  HTMLClassAdd (e, cls2);
}

function updateActionsChecked (form, target, name)
{
  for (var i = 0, len = form.elements.length; i < len; i++)
  {
    var e = form.elements[i];

    if (e.type == 'checkbox' && e.checked && e.name.indexOf(name) != -1)
      return HTMLClassRemove (target, 'disabled');
  }

  HTMLClassAdd (target, 'disabled');
}

function checkUncheckAll (e, name, checkedText, uncheckedText)
{
  var r = document.querySelectorAll ('input[name^="'+name+'"]');
  var v = e.getAttribute ('data-checked');

  for (var i = 0, len = r.length; i < len; i++)
    r[i].checked = (v == 0);

  if (v == 1)
  {
    e.setAttribute ('data-checked', 0);
    HTMLClassReplace (e.querySelector('i'), 'fa-square-o', 'fa-check-square-o');
    e.querySelector('span').innerText = checkedText;
  }
  else
  {
    e.setAttribute ('data-checked', 1);
    HTMLClassReplace (e.querySelector('i'), 'fa-check-square-o', 'fa-square-o');
    e.querySelector('span').innerText = uncheckedText;
  }
}

function displayMsg (msg, type, title)
{
  var e = document.getElementById ('alert-msg');

  e.innerHTML =
    "<button class='close' onclick='document.getElementById(\"alert-msg\").style.display=\"none\"'>&times;</button>"+((title)?"<strong>"+title+"</strong>":'')+"<p>"+msg+"</p>";

  e.className = 'alert alert-'+((type)?type:'info');
  e.style.display = 'block';
}


function translate_console_open (lang2)
{
  lang2 = (lang2) ? lang2.replace (/^([a-z]+).*/i, "$1") : '';

  window.open(
    'translate_popup.cgi?lang2='+lang2,
    'translate_popup',
    'toolbar=no,location=no,directories=no,status=no,scrollbars=yes,' +
    'resizable=no,copyhistory=no,width=600,height=400'
  );
}
